### PR TITLE
Lensfunpy

### PR DESCRIPTION
**Pull request by mwilson8**
Checklist

- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [ ] Source is from official source
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Build number is 0
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by mwilson8** [comment]
Getting the error ```
conda.exceptions.ResolvePackageNotFound: 
  - lensfun==0.3.95=0
```

@conda-forge/help-c-cpp @conda-forge/help-python-c  @conda-forge/staged-recipes  I'm not sure the cause for the error, as `lensfun` is listed as a dependency and is currently in the forge, so I'm unsure of which team to ask

**Comment by beckermr** [comment]
Frack. This is my fault. The run export in the `lensfun` package is wrong. I don't get why, but my guess is that the hash depends on the run_export contents which needs the package hash. I will make a PR on the upstream repo to fix. Sit tight!

**Comment by beckermr** [comment]
Mmmmk. Merge this one when it passes and try again. 

https://github.com/conda-forge/lensfun-feedstock/pull/1

**Comment by beckermr** [comment]
OK. Wait about 2 hours and then retry this build. It takes a little while for new packages to hit the index.

**Comment by beckermr** [comment]
I will look into this further tonight.

**Comment by beckermr** [comment]
These build errors are not due to the lensfun package missing a run export. It looks like something else. There is a git submodule that is being pulled but failing and a compilation issue as well. Check out the logs https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=95521&view=logs

**Comment by mbargull** [comment]
```
lensfunpy/_lensfun.c:8158:16: error: too few arguments to function 'lf_lens_interpolate_vignetting'
```
https://github.com/letmaik/lensfunpy/blob/v1.6.1/lensfunpy/_lensfun.pyx#L164
https://github.com/lensfun/lensfun/blob/v0.3.95/include/lensfun/lensfun.h.in#L1624
https://github.com/lensfun/lensfun/commit/f4327669dd8bfc3b1b1214ae6c0460d9a2889775

```
Breaking changes:

* C interface: all lf_lens_interpolate_...() functions now require an additional crop parameter
```

=> `lensfunpy` needs to be updated for the newer `lensfun` version. (or `lensfun<0.3.95` would need to be packaged)

**Comment by mwilson8** [comment]
@mbargull @beckermr  per the (issue from the author here)[https://github.com/letmaik/lensfunpy/issues/16]

It doesn't seem like `lensfunpy` will be updated

**Comment by oblute** [comment]
@mbargull @beckermr i'm going to try packaging `lensfun v0.3.2` to remedy this issue because @mwilson8 referenced looks like `lensfunpy` may not be updated for awhile

**Comment by beckermr** [comment]
Sounds good. You can make a pr against the existing feedstock.

**Comment by mwilson8** [comment]
@mbargull @beckermr there seems to be a problem in the `setup.py` on mac, where it's (I'm guessing) trying to rebuild `lensfun` from scratch??... I'm a little out of my depth on what we could modify in order to have it use conda's dependency management
[build](https://dev.azure.com/oliviablute/to24-condify/_build/results?buildId=441)

**Comment by mbargull** [comment]
We could patch https://github.com/letmaik/lensfunpy/blob/v1.6.1/setup.py#L222
like so:
```diff
-needsCompile = any(s in cmdline for s in ['install', 'bdist', 'build_ext', 'nosetests'])
+needsCompile = False
```
maybe?

**Comment by beckermr** [comment]
I was in the middle of suggesting a patch when @mbargull beat me to it. Yes this is the right way to go.

**Comment by beckermr** [comment]
See this: https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#patches

File: recipes/lensfunpy/meta.yaml [hunk 1]
```diff
@@ -18,17 +19,25 @@ build:
...
   run:
     - python
-    
+    - cython
+    - nose
+    - wheel >=0.25
+    - sphinx
+    - sphinx_rtd_theme
+    - scipy
+    - lensfun ==0.3.2
```
Comment (beckermr) [inline recipes/lensfunpy/meta.yaml]: Why the exact pin here and not above?

Comment (mwilson8) [reply recipes/lensfunpy/meta.yaml]: copied the run deps part from Olivia's meta, where as the others were already there from mine. Do you have a preference as to which to use?

**Comment by mwilson8** [comment]
@beckermr we're confused why this only fails on osx.. could you take a look when you get a chance

**Comment by beckermr** [comment]
You may need something like `-I${PREFIX}/include` in the CFLAGS, but let's see if the c compiler helps.

**Comment by beckermr** [comment]
The compilers use `-isystem` for the PREFIX include paths. IDK if that means clang is expecting `#include <lensfun.h>` as opposed to ``#include "lensfun.h"`.

**Comment by mwilson8** [comment]
> The compilers use `-isystem` for the PREFIX include paths. IDK if that means clang is expecting `#include <lensfun.h>` as opposed to ``#include "lensfun.h"`.

I had briefly considered that, but would that mean we would have to patch all the files.. or is there a flag we can add to work around this? Also unsure, however

**Comment by isuruf** [comment]
There's an osx+win switch at https://github.com/letmaik/lensfunpy/blob/master/setup.py#L67-L74.

**Comment by beckermr** [comment]
It looks like GCC does both which explains why the linux builds are fine

```
Directories specified with -I, -isystem, or -idirafter apply to lookup for both the #include "file" and #include <file> directives.
```

**Comment by oblute** [comment]
> There's an osx+win switch at https://github.com/letmaik/lensfunpy/blob/master/setup.py#L67-L74.

how does that relate specifically to the failure we are seeing?

**Comment by beckermr** [comment]
@mwilson8 If you add the `-I` version to the CFLAGS, it should work without changing the source code.

**Comment by isuruf** [comment]
> how does that relate specifically to the failure we are seeing?

On linux the correct directory is added which is `$PREFIX/include/lensfun`. In OSX, a wrong directory is added, `/private/tmp/pip-req-build-iyrvre73/external/lensfun/cmake_build/install/include/lensfun`

**Comment by beckermr** [comment]
Ahhhhh good catch @isuruf ! Thank you!

**Comment by mwilson8** [comment]
> > how does that relate specifically to the failure we are seeing?
> 
> On linux the correct directory is added which is `$PREFIX/include/lensfun`. In OSX, a wrong directory is added, `/private/tmp/pip-req-build-iyrvre73/external/lensfun/cmake_build/install/include/lensfun`

should we patch this file to change the directory to match the linux structure

**Comment by beckermr** [comment]
Patch the setup.py

**Comment by mwilson8** [comment]
> Patch the setup.py

This patch seems like it is a little out of my depth.. how much are we needing to change? Could we just remove the win/osx if statement & let all cases use `use_pkg_config()`?

**Comment by beckermr** [comment]
Probably easiest to just set the value correctly in the if statement as opposed to trying to use pkg_config which may not work. Something like `os.path.join(or.environ['PREFIX'], 'include', 'lensfun')` is the right value for the include path.

**Comment by isuruf** [comment]
It's good to just remove that condition for osx and let it use pkg-config

**Comment by isuruf** [comment]
You also need to add `pkg-config` to `build` requirements.

**Comment by beckermr** [comment]
Ahhh ok. I'm gonna let @isuruf take over here! ;p

**Comment by mwilson8** [comment]
@isuruf there's a function called `use_pkg_config()` but you're saying we can use the conda-package pkg-config?

**Comment by isuruf** [comment]
> @isuruf there's a function called use_pkg_config() but you're saying we can use the conda-package pkg-config?

No, what I'm saying is for `use_pkg_config()` python function to work an executable called `pkg-config` needs to exist. (See https://github.com/letmaik/lensfunpy/blob/master/setup.py#L35-L38) This executable is in `pkg-config` conda package.

**Comment by mwilson8** [comment]
@isuruf please review

File: recipes/lensfunpy/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+build:
+  number: 0
+  script: {{ PYTHON }} -m pip install . -vv 
+  skip: True  # [win or py<27]
+   
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - lensfun <0.3.95
```
Comment (isuruf) [inline recipes/lensfunpy/meta.yaml]: This should only be in `host`

File: recipes/lensfunpy/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    - pip
+    - python
+    - lensfun <0.3.95
+    - numpy
+    
+  run:
+    - python
+    - cython
+    - nose
+    - wheel >=0.25
```
Comment (isuruf) [inline recipes/lensfunpy/meta.yaml]: Why do you need `wheel` at runtime?

File: recipes/lensfunpy/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    - sphinx
+    - sphinx_rtd_theme
```
Comment (isuruf) [inline recipes/lensfunpy/meta.yaml]: sphinx is for docs which is not needed at runtime

File: recipes/lensfunpy/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  host:
+    - pip
+    - python
+    - lensfun <0.3.95
+    - numpy
+    
+  run:
+    - python
+    - cython
+    - nose
```
Comment (isuruf) [inline recipes/lensfunpy/meta.yaml]: This is needed only for testing. No need at runtime.

File: recipes/lensfunpy/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - lensfun <0.3.95
+    - pkg-config        
+  host:
+    - pip
+    - python
+    - lensfun <0.3.95
+    - numpy
```
Comment (isuruf) [inline recipes/lensfunpy/meta.yaml]: See https://conda-forge.org/docs/maintainer/knowledge_base.html#building-against-numpy

**Comment by mwilson8** [comment]
@isuruf please re-review changes

File: recipes/lensfunpy/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    - {{ compiler('cxx') }}
+    - pkg-config        
+  host:
+    - pip
+    - python
+    - lensfun <0.3.95
+    - numpy
+    - sphinx
+    - wheel >=0.25
+    - enum34  # [py<34]
```
Comment (isuruf) [inline recipes/lensfunpy/meta.yaml]: Do you need the 3 packages above?

File: recipes/lensfunpy/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  run:
+    - {{ pin_compatible('numpy') }}
+    - python
+    - cython
+    - scipy
+    - lensfun <0.3.95
```
Comment (isuruf) [inline recipes/lensfunpy/meta.yaml]:
```suggestion
    - lensfun <0.3.95
    - {{ pin_compatible("lensfun") }}
```

File: recipes/lensfunpy/meta.yaml (hunk 7 above)
Comment (mwilson8) [reply recipes/lensfunpy/meta.yaml]: They were listed in a `dev-requirements.txt` file so we weren't sure

**Comment by mwilson8** [comment]
@isuruf ready for review

File: recipes/lensfunpy/meta.yaml (hunk 7 above)
Comment (mwilson8) [reply recipes/lensfunpy/meta.yaml]: But enum, yes: https://github.com/letmaik/lensfunpy/blob/master/lensfunpy/_lensfun.pyx#L9

File: recipes/lensfunpy/meta.yaml [hunk 9]
```diff
@@ -26,15 +26,13 @@ requirements:
...
-    - sphinx
-    - wheel >=0.25
     - enum34  # [py<34]
   run:
     - {{ pin_compatible('numpy') }}
     - python
     - cython
     - scipy
-    - lensfun <0.3.95
+    - {{ pin_compatible('lensfun') }}
```
[the remark this answers is not in the document]
Comment (isuruf) [reply recipes/lensfunpy/meta.yaml]: Both are needed IMO.

File: recipes/lensfunpy/meta.yaml (hunk 7 above)
Comment (isuruf) [reply recipes/lensfunpy/meta.yaml]: That's needed at runtime, so no need for it to be in `host`. Only `run` is enough

File: recipes/lensfunpy/LICENSE [hunk 10]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Comment (isuruf) [inline recipes/lensfunpy/LICENSE]: Can you open an issue upstream about including the license in the sdist?

**Comment by mwilson8** [comment]
https://github.com/letmaik/lensfunpy/issues/23
But with the author not having commited, commented in over a year the chances are likely low

**Pull request merged**